### PR TITLE
mark multimedia export as location safe

### PR DESCRIPTION
**Pull request by jingcheng16**
## Technical Summary

Ticket: https://dimagi.atlassian.net/browse/SAAS-12925

Location-restricted user is not able to export multimedia because the view to export multimedia doesn't have @location_safe decorator. The view to export multimedia will use ExpandedMobileWorkerFilter to limit its access only to the assigned locations.

Context about `@location_safe` decorator [here](https://github.com/dimagi/commcare-hq/blob/d5ccaf4807da801ac58548b86d5104e08899bfd4/corehq/apps/locations/permissions.py#L65-L71).

This ticket is similar to https://dimagi.atlassian.net/browse/HI-208

## Feature Flag

No

### Safety story

Test on staging: setting up a location restricted web user who is able to download multimedia, and the multimedia is only restricted to the location that he have access to.

### QA Plan

QA: https://dimagi.atlassian.net/browse/QA-7285

QA plan: make sure location restricted user can only export multimedia from their assigned locations.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by mkangia** [comment]
How is multimedia restricted to locations? or how do we find out which multimedia is accessible to a location?
I assume multimedia here is app multimedia, which, I would think of being location independent. 

> The view to export multimedia will use ExpandedMobileWorkerFilter to limit its access only to the assigned locations.

Is this change yet to be made?

**Comment by mkangia** [comment]
> It is multimedia submitted by mobile workers in a form submission, for example the audio, the photo...

Okay :+1: 

and, same question as Daniel about the view. I don't understand how its related.

**Comment by jingcheng16** [comment]
@millerdev @mkangia 
In `prepare_form_multidedia`, it will first determine whether it is [form_or_case](https://github.com/dimagi/commcare-hq/blob/6f3ee2448082314275a597e165dd8ec78668128f/corehq/apps/export/views/download.py#L453), multimedia is only available in form export, so it will be form. Then we [decide the view helper here](https://github.com/dimagi/commcare-hq/blob/6f3ee2448082314275a597e165dd8ec78668128f/corehq/apps/export/views/download.py#L458). We will use [FormDownloadExportViewHelper](https://github.com/dimagi/commcare-hq/blob/6f3ee2448082314275a597e165dd8ec78668128f/corehq/apps/export/views/download.py#L91-L92) because the model is form. This view helper use [EmwfFilterFormExport](https://github.com/dimagi/commcare-hq/blob/6f3ee2448082314275a597e165dd8ec78668128f/corehq/apps/export/views/download.py#L129)

**Comment by millerdev** [comment]
I'm still finding it difficult to evaluate how `ExpandedMobileWorkerFilter` is connected to `prepare_form_multidedia`, and so I'm hesitant to approve without being confident that this is a safe change. Maybe @esoergel or @orangejenny has more context?

**Comment by jingcheng16** [comment]
@millerdev No problem! Thank you Daniel for looking into it! I just updated the comment. Can you let me know which part make you feel difficult? Maybe I can modify my response to make it more clear. I agree it is hard to follow because there are too many layers...

**Comment by esoergel** [comment]
good find, Jing - that does seem like it's probably location-safe.  This seems way more involved than it needs to be, so frustrating to work with.  Anyways, do you have a sense of how testable this is?  It'd be super helpful to be able to have an automated showing that location-restricted users can't access forms outside their scope, especially given the complexity -that'd help prevent any future refactors from accidentally removing that layer of safety.

**Comment by millerdev** [comment]
+1. A test is a great idea.

**Comment by jingcheng16** [comment]
@millerdev @esoergel 
I spent a few hours yesterday trying to write tests for it, but cannot setup it correctly. 
```
url = reverse('prepare_form_multimedia, 'self.domain.name')
return self.client.post(url, post_data)
```
When I try to post to the `prepare_form_multimedia`  it always give me 404...

I find it is hard for me to write the test, because I only figured out it basically follows the pattern that is used by form export, but there are a lot of things I don't understand, maybe this is the reason that I'm not able to write a test for it on my own...

Can we move forward without writing a test?

1) It use the same logic as form export, there is no test for form export, why a test for form multimedia export is that necessary?
2) I have tested my change on staging. I can put it through QA as well.

If I have time this quarter, I would be willing to dive into how to write tests for this. But I'm also struggled to complete what I commit to in Q4, and becoming anxious when I put too much time on this ticket.

**Review by millerdev — APPROVED** [review]
> It use the same logic as form export

This is the most convincing to me. Approving due to urgency, but it would be nice if we could eventually make a test that confirms that `@location_safe` does not allow a user to download multimedia they should not have permission to download.

**Comment by jingcheng16** [comment]
Thank you Daniel. 
I just want to clarify, we need to first make sure the view function does not allow a user to download multimedia outside of their assigned locations. Then we know it is location safe, then we add `location_safe` decorator. The decorator itself won't ensure anything.
